File: matlab/Figure9_3and4.m
% Author: Kenji Kashima
% Date  : 2025/09/01

rng(2);

% shared parameters
k_update = 50;
Npath    = 20;

% Figure 9.3(a)
iter_Gain = 1; sigma = 2;
[x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain] = RL4LQR(sigma, iter_Gain, k_update, Npath);
plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain, '3a');

% Figure 9.3(b)
sigma = 10;
[x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain] = RL4LQR(sigma, iter_Gain, k_update, Npath);
plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain, '3b');

% Figure 9.4
iter_Gain = 5;
[x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain] = RL4LQR(sigma, iter_Gain, k_update, Npath);
plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain, '4');


% =====================================================================
function [x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain] = RL4LQR(sigma, iter_Gain, k_update, Npath)
% RL for LQR with TD+RLS (Algorithm 9.x equivalent)

beta = 0.95;

A = [0.80 0.90 0.86;
     0.30 0.25 1.00;
     0.10 0.55 0.50];
B = [1; 0; 0];

x_dim = size(A,1);
u_dim = size(B,2);
n     = x_dim + u_dim;
p_dim = n*(n+1)/2+1;

Q = eye(x_dim);
R = eye(u_dim);

% discounted optimal LQR (comparison)
[K_opt, ~, ~] = dlqr( sqrt(beta)*A, sqrt(beta)*B, Q, R );   % 1x3

K = [4.1100, 11.7519, 19.2184];          % initial gain (1x3)
x = zeros(x_dim,1);                      % initial state

x_norm_hist_p = [];
K_err_hist_p  = [];
Ups_err_hist_p= [];

for j = 1:Npath

    x_hist        = x;
    K_err_hist    = zeros(1, iter_Gain*k_update);
    Ups_err_hist  = zeros(1, iter_Gain*k_update);

    isUnstable = false;
    t_idx = 0;

    for it = 1:iter_Gain
        % --- true Pi, Ups_true for current K ---
        Ak = A - B*K;                        
        G  = [eye(x_dim); -K];
        QK = G.' * G;                        

        % Pi solves: Pi = beta*Ak*Pi*Ak' + QK
        Pi = dlyap( sqrt(beta)*Ak', QK );

        Ups_true = beta * [A, B].' * Pi * [A, B] + blkdiag(Q, R); 

        % --- TD + RLS block for k_update steps ---
        p     = zeros(p_dim,1);
        Sigma = 10*eye(p_dim);

        for k = 1:k_update
            t_idx = t_idx + 1;

            % current errors
            K_err_hist(t_idx)   = norm(K - K_opt, 'fro') / norm(K_opt, 'fro');
            Ups_est             = p_to_Ups(p, n);
            Ups_err_hist(t_idx) = norm(Ups_est - Ups_true, 'fro') / norm(Ups_true, 'fro');

            % exploration input and cost
            u    = -K*x + sigma*randn(u_dim,1);
            cost = x.'*Q*x + u.'*R*u;

            % TD update
            phi_pre = phi(x, u);             % (p_dim x 1)
            x       = A*x + B*u;             % state step
            x_hist  = [x_hist, x];

            q       = phi_pre - beta * phi(x, -K*x);

            denom   = 1 + q.'*Sigma*q;
            p       = p + (Sigma*q) * ((cost - q.'*p)/denom);
            Sigma   = Sigma - (Sigma*(q*q.')*Sigma)/denom;
            % Sigma = 0.5*(Sigma+Sigma.');   % 対称化したい場合は解除
        end

        % --- gain improvement: K = Uuu \ Uux  (eq. 9.24) ---
        Ups = p_to_Ups(p, n);                % 4x4
        Uuu = Ups(x_dim+1:end, x_dim+1:end); % 1x1
        Uux = Ups(x_dim+1:end, 1:x_dim);     % 1x3
        K   = (Uuu \ Uux);                   % 1x3 (row vector)

        % stability check
        if max(abs(eig(A - B*K))) > 1
            isUnstable = true;
            break;
        end
    end

    if ~isUnstable
        x_norm_hist_p = [x_norm_hist_p; vecnorm(x_hist,2,1)]; %#ok<AGROW>
        K_err_hist_p  = [K_err_hist_p;  K_err_hist];          %#ok<AGROW>
        Ups_err_hist_p= [Ups_err_hist_p;Ups_err_hist];        %#ok<AGROW>
    end
end
end


% =====================================================================
function phi_vec = phi(x, u)
% Feature vector for Q-function: upper-triangular vectorization of [x;u][x;u]^T
xu = [x; u];                     % n x 1
H  = xu * xu.';                  % n x n
n  = size(H,1);
phi_vec = zeros(n*(n+1)/2+1, 1);
idx = 0;
for i = 1:n
    m = n - i + 1;
    phi_vec(idx+1:idx+m) = H(i, i:end).';
    idx = idx + m;
end
phi_vec(end) = 1;   % constant term
end

% =====================================================================
function Ups = p_to_Ups(p, n)
% Recover symmetric matrix Ups from its upper-triangular vectorization
p   = p(1 : n*(n+1)/2);
Ups = zeros(n,n);
idx = 0;
for i = 1:n
    m = n - i + 1;
    Ups(i, i:end) = p(idx+1:idx+m).';
    idx = idx + m;
end
Ups = 0.5*(Ups + Ups.');
end

% =====================================================================
function plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain, label)
gray = [0.7 0.7 0.7];
T = size(x_norm_hist_p,2);

fig = figure('Name',['Figure 9.' label]);
tiledlayout(2,1,'TileSpacing','compact','Padding','compact');

% ||x|| history
nexttile; hold on; grid on;
plot(x_norm_hist_p.', 'Color', gray);
plot(mean(x_norm_hist_p,1), 'k', 'LineWidth', 2);
xlim([1, T]); yline(0,'k--','LineWidth',1);
ylabel('State norm');

% Q-function error (+ Gain error if iter_Gain>1)
nexttile; hold on; grid on;
plot(Ups_err_hist_p.', 'Color', gray);
plot(mean(Ups_err_hist_p,1), 'k', 'LineWidth', 1.5);
ylabel('Q function error'); xlim([1, T]); yline(0,'k--','LineWidth',1);

if iter_Gain > 1
    yyaxis right;
    plot(mean(K_err_hist_p,1), 'r', 'LineWidth', 2);
    ylabel('Gain error'); set(gca,'YColor',[1 0 0]);
end

xlabel('$k$', 'Interpreter','latex');
set(fig, 'Position',[120 120 700 520]);

end

File: python/Figure9_3and4.py
# Author: Kenji Kashima
# Date  : 2025/09/01
# Note  : pip install control

import numpy as np
import matplotlib.pyplot as plt
from scipy.linalg import solve_discrete_lyapunov as dlyap
import control
import config

np.random.seed(2)

# Function for the RL4LQR algorithm
def RL4LQR(sigma, iter_Gain, k_update, Npath):
    """
    Simulate RL for LQR

    Parameters
    ----------
    sigma           : scale of exploration noise
    iter_Gain       : number of gain updates
    k_update        : time steps for one gain update (k_update)
    Npath           : number of sample paths

    Returns
    -------
    x_norm_hist_p   : history of state norms
    K_err_hist_p    : history of control gain errors
    Ups_err_hist_p  : history of H errors
    iter_Gain       : number of gain updates
    """

    beta = 0.95  # discount rate
    A = np.array([[0.80, 0.90, 0.86],
                  [0.30, 0.25, 1.00],
                  [0.10, 0.55, 0.50]])  # A-matrix
    B = np.array([[1], [0], [0]])  # B-matrix

    x_dim, u_dim = B.shape  # state & input dimension
    p_dim = (x_dim + u_dim) * (x_dim + u_dim + 1) // 2  + 1 # size of p (integer)

    Q = np.eye(x_dim)  # cost x'Qx
    R = np.eye(u_dim)  # cost u'Ru

    # Optimal gain for comparison
    K_opt, _, _ = control.dlqr(np.sqrt(beta) * A, np.sqrt(beta) * B, Q, R)

    K_ini = np.array([[4.1100, 11.7519, 19.2184]])
    IsUnstable = False
    x_ini = np.zeros((x_dim, 1))  # Initial state of dynamics

    x_norm_hist_p = []
    Ups_err_hist_p = []
    K_err_hist_p = []

    for j in range(Npath):
        
        K_err_hist = []
        Ups_err_hist = []

        p = np.zeros(p_dim)

        K = K_ini
        x = x_ini
        x_hist = x
        for _ in range(iter_Gain):
            # Compute Ups_true for current K for comparison
            Pi = dlyap(np.sqrt(beta) * (A - B @ K).T, np.hstack([np.eye(x_dim), -K.T]) @ np.vstack([np.eye(x_dim), -K]))  # Solve Lyapunov equation (9.22)
            Ups_true = beta * np.hstack([A, B]).T @ Pi @ np.hstack([A, B]) + np.block([[Q, np.zeros((x_dim, u_dim))], [np.zeros((u_dim, x_dim)), R]])   # eq.(9.34)

            # TD learning 

            # initialization of RLS
            p = np.zeros((p_dim,1))     
            Sigma = np.eye(p_dim) * 10

            for k in range(k_update):
               
                K_err_hist.append(np.linalg.norm(K - K_opt) / np.linalg.norm(K_opt))
                Ups_err_hist.append(np.linalg.norm(p_to_Ups(p, x_dim + u_dim) - Ups_true) / np.linalg.norm(Ups_true))

                u = -K @ x + np.random.randn(u_dim, 1) * sigma  # tentative SF + exploration noise
                cost = x.T @ Q @ x + u.T @ R @ u    # stage cost

                phi_pre = phi(x, u)
                x = A @ x  + B @ u                  # One-step simulation
                x_hist= np.hstack((x_hist,x))

                q = phi_pre - beta * phi(x, -K @ x) # line 1 
                denom = 1 + q.T @ Sigma @ q
                p = p + (Sigma @ q @ (cost - q.T @ p)) / denom  # line 2
                Sigma = Sigma - (Sigma @ q @ q.T @ Sigma) / denom   # line 3

            Ups = p_to_Ups(p, x_dim + u_dim)
            K = np.linalg.solve(Ups[x_dim:, x_dim:], Ups[x_dim:, :x_dim])   # eq.(9.24)

            if max(abs(np.linalg.eigvals(A - np.dot(B, K)))) > 1:
                IsUnstable = True
                break

        if not IsUnstable:
            x_norm_hist_p.append(np.linalg.norm(x_hist, axis=0))
            K_err_hist_p.append(K_err_hist)
            Ups_err_hist_p.append(Ups_err_hist)
        else:
            IsUnstable = False
    return (x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain)
    

# Function to compute feature vector phi from x and u; eq.(9.35)
def phi(x, u):
    xu = np.vstack([x, u])              # (n,1)
    H  = xu @ xu.T                      # (n,n)
    phi_vector = np.hstack([H[i, i:] for i in range(H.shape[0])]).reshape(-1, 1)
    return np.concatenate([phi_vector, [[1.0]]])

# Function to convert p vector to Upsilon matrix; eq.(9.35)
def p_to_Ups(p, n):
    p = np.asarray(p).reshape(-1)  # 1D
    p = p[:n*(n+1)//2]                      # drop trailing bias

    H = np.zeros((n, n))
    idx = 0
    for i in range(n):
        H[i, i:] = p[idx:idx + n - i]
        idx += n - i
    return (H + H.T) / 2

# Function to plot results
def plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain, label=None):
    grayColor = [0.7, 0.7, 0.7]
    figsize = config.global_config(type= 1)
    fig, ax = plt.subplots(2, 1, figsize=figsize)
    
    # Plot x_norm_hist_p
    ax[0].plot(np.array(x_norm_hist_p).T, color=grayColor)
    ax[0].plot(np.mean(x_norm_hist_p, axis=0), linewidth=2, color='black')
    ax[0].set_xlim([0,len(np.array(x_norm_hist_p).T)-1])
    ax[0].axhline(y=0, color='black', linestyle='--', linewidth=1)

    # Plot Ups_err_hist_p
    ax[1].plot(np.array(Ups_err_hist_p).T, color=grayColor)
    ax[1].plot(np.mean(Ups_err_hist_p, axis=0), linewidth=1.5, color='black')
    ax[1].set_ylabel('Q function error')
    ax[1].set_xlim([0,len(np.array(Ups_err_hist_p).T)-1])
    ax[1].axhline(y=0, color='black', linestyle='--', linewidth=1)

    plt.xlabel(r'$k$')
    # Plot K_err_hist_p if iter_Gain > 1
    if iter_Gain > 1:
        ax2 = ax[1].twinx()  # Create a twin y-axis
        ax2.plot(np.mean(K_err_hist_p, axis=0), linewidth=2, color='red')
        ax2.set_ylabel('Gain error', color='red')

    plt.tight_layout()
    
    if label is not None:
        plt.savefig("./Figure9_{}.pdf".format(label))
    plt.show()

if __name__ == '__main__':
    # Running the RL4LQR for different scenarios
    k_update = 50
    Npath = 20

    # Figure 9.3(a)
    iter_Gain = 1
    sigma = 2
    x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain = RL4LQR(sigma, iter_Gain, k_update, Npath)
    plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain,label="3a")
    
    # Figure 9.3(b)
    sigma = 10
    x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain = RL4LQR(sigma, iter_Gain, k_update, Npath)
    plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain,label="3b")

    # Figure 9.4
    iter_Gain = 5
    x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain = RL4LQR(sigma, iter_Gain, k_update, Npath)
    plot_results(x_norm_hist_p, K_err_hist_p, Ups_err_hist_p, iter_Gain, label="4")

